python/lsst/pipe/base/pipelineTask.py

```diff
@@ -309,24 +364,47 @@ def runQuantum(self, quantum, butler):
                 inputs[key] = [butler.get(dataRef.datasetType.name, dataRef.dataId)
                                for dataRef in dataRefs]
 
-        # lists of units for output datasets
-        outUnits = {}
+        # lists of DataRefs/DataIds for output datasets
+        outputDataRefs = {}
+        outputDataIds = {}
         for key, value in self.config.items():
             if isinstance(value, OutputDatasetConfig):
                 dataRefs = quantum.outputs[value.name]
-                outUnits[key] = [dataRef.dataId for dataRef in dataRefs]
+                outputDataRefs[key] = dataRefs
+                outputDataIds[key] = [dataRef.dataId for dataRef in dataRefs]
 
         # call run method with keyword arguments
-        struct = self.run(**inputs, **outUnits)
+        struct = self.runWithOutputs(inputs, outputDataIds)
+
+        # store produced ouput data
+        self._saveStruct(struct, outputDataRefs, butler)
+
+    def _saveStruct(self, struct, outputDataRefs, butler):
```

Comment:
I don't see a good reason to treat `_saveStruct` and `runWithOutputs` differently w.r.t. leading underscores - in C++, I'd have called them both (non-pure) virtual protected, and while I don't have a strong opinion about whether that merits an underscore in Python, I think they should be treated consistently.

Reply:
In my (twisted) mental model `_saveStruct()` is indeed protected (or even private) and not to be called by clients directly, while `runWithOutputs()` is public because it is the same as `run()`, only better :) Anyways, I'll make thing uniform-no-underscores.